ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts

```diff
@@ -0,0 +1,193 @@
+   *
+   * @see {@link
+   * https://aptos.dev/integration/creating-a-signed-transaction/ | Creating a Signed Transaction}
+   *
+   * @param publicKeys A list of public keys
+   * @param threshold At least "threshold" signatures must be valid
+   */
+  constructor(args: { publicKeys: PublicKey[]; threshold: number }) {
+    const { publicKeys, threshold } = args;
+
+    // Validate number of public keys
+    if (publicKeys.length > MultiPublicKey.MAX_KEYS || publicKeys.length < MultiPublicKey.MIN_KEYS) {
+      throw new Error(
+        `Must have between ${MultiPublicKey.MIN_KEYS} and ${MultiPublicKey.MAX_KEYS} public keys, inclusive`,
+      );
+    }
+
+    // Validate threshold: must be between 1 and the number of public keys, inclusive
+    if (threshold < MultiPublicKey.MIN_THRESHOLD || threshold > publicKeys.length) {
+      throw new Error(`Threshold must be between ${MultiPublicKey.MIN_THRESHOLD} and ${publicKeys.length}, inclusive`);
+    }
+
+    this.publicKeys = publicKeys;
+    this.threshold = threshold;
+  }
+
+  /**
+   * Converts a PublicKeys into Uint8Array (bytes) with: bytes = p1_bytes | ... | pn_bytes | threshold
+   */
+  toUint8Array(): Uint8Array {
+    const bytes = new Uint8Array(this.publicKeys.length * PublicKey.LENGTH + 1);
+    this.publicKeys.forEach((k: PublicKey, i: number) => {
+      bytes.set(k.toUint8Array(), i * PublicKey.LENGTH);
+    });
+
+    bytes[this.publicKeys.length * PublicKey.LENGTH] = this.threshold;
+
+    return bytes;
+  }
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeBytes(this.toUint8Array());
+  }
+
+  static deserialize(deserializer: Deserializer): MultiPublicKey {
+    const bytes = deserializer.deserializeBytes();
+    const threshold = bytes[bytes.length - 1];
+
+    const keys: PublicKey[] = [];
+
+    for (let i = 0; i < bytes.length - 1; i += PublicKey.LENGTH) {
+      const begin = i;
+      keys.push(new PublicKey({ hexInput: bytes.subarray(begin, begin + PublicKey.LENGTH) }));
+    }
+    return new MultiPublicKey({ publicKeys: keys, threshold });
+  }
+}
+
+export class MultiSignature {
+  // Maximum number of signatures supported
+  static MAX_SIGNATURES_SUPPORTED = 32;
+
+  // Bitmap length
+  static BITMAP_LEN: number = 4;
+
+  public readonly signatures: Signature[];
+
+  public readonly bitmap: Uint8Array;
+
+  /**
+   * Signature for a K-of-N multisig transaction.
+   *
+   * @see {@link
+   * https://aptos.dev/integration/creating-a-signed-transaction/#multisignature-transactions | Creating a Signed Transaction}
+   *
+   * @param args.signatures A list of signatures
+   * @param args.bitmap 4 bytes, at most 32 signatures are supported. If Nth bit value is `1`, the Nth
+   * signature should be provided in `signatures`. Bits are read from left to right
+   */
+  constructor(args: { signatures: Signature[]; bitmap: Uint8Array }) {
+    const { signatures, bitmap } = args;
+    if (bitmap.length !== MultiSignature.BITMAP_LEN) {
```

Comment:
do we also want to check args.signatures length?

Reply:
Good catch!